Updated linalg.lstsq with NumPy compatible kwarg rcond

IvanYashchuk opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #54725 Added cuBLAS path for torch.linalg.lstsq
* #54724 Replaced _lstsq_helper with internal dispatch
* **#54723 Updated linalg.lstsq with NumPy compatible kwarg rcond**
* #54722 Modified batched residuals return of torch.linalg.lstsq

Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

Differential Revision: [D27993741](https://our.internmc.facebook.com/intern/diff/D27993741)

IvanYashchuk committed:
Updated linalg.lstsq with NumPy compatible kwarg rcond

Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk commented:
This PR is introducing backward-incompatible changes to `aten::linalg_lstsq`, but lstsq was merged only recently (on March 17). SciPy uses `cond` kwarg for specifying the tolerances while NumPy uses `rcond`, we should use `rcond` as well for NumPy compatibility.
I added `aten::linalg_lstsq` to allow_list in `check_backward_compatibility.py` with the end date in May.

In the future, I think a better name `rtol` should be added and promoted (Ref. https://github.com/pytorch/pytorch/issues/54151).

IvanYashchuk commented on test/test_linalg.py:
```diff
@@ -199,29 +199,22 @@ def select_if_not_empty(t, i):
+                    return np.linalg.lstsq(a, b, rcond=rcond)
                 check_correctness_ref(a, b, res, numpy_ref)
 
-        def check_ranks(a, ranks, cond=1e-7):
-            ranks2 = torch.matrix_rank(a, tol=cond)
-            self.assertEqual(ranks, ranks2)
-
-        def check_singular_values(a, sv):
-            sv2 = a.svd()[1]
-            self.assertEqual(sv, sv2)
```
Rank and singular values are already checked in `check_correctness_scipy` and `check_correctness_numpy`.

mruberry commented on test/test_linalg.py:
```diff
@@ -232,32 +225,35 @@ def check_singular_values(a, sv):
+        rconds = (None, True, -1)
+
+        for batch, matrix_size, driver, rcond in itertools.product(batches, matrix_sizes, drivers, rconds):
+            if rcond and driver != 'gels':
```
These two if statements seem like they could be written in a more straightforward way? Maybe I'm reading this incorrectly

IvanYashchuk replied:
Looking at it again I see that is indeed confusing and not correct. I refined the if statements:
```python
# keep the rcond value if it is None or -1, set the driver specific value if it is True
if rcond and rcond != -1:
    if driver in ('gelss', 'gelsd'):
        # SVD based algorithm; set to zero roughly half of all the singular values
        rcond = 1.0
    else:
        # driver == 'gelsy'
        # QR based algorithm; setting the value too high might lead to non-unique solutions and flaky tests
        rcond = 1e-3

# specifying rcond value has no effect for gels driver so no need to run the tests again
if driver == 'gels' and rcond != None:
    continue
```

mruberry commented on test/test_linalg.py:
```diff
@@ -232,32 +225,35 @@ def check_singular_values(a, sv):
+            check_correctness_scipy(a, b, res, driver, rcond)
+            check_correctness_numpy(a, b, res, driver, rcond)
+
+            # gels driver is not checked comparing to NumPy or SciPy
```
I'm not sure I follow -- aren't the calls to check_correctness{scipy, numpy} above going to happen before this if? Or am I reading this incorrectly?

IvanYashchuk replied:
`check_correctness_numpy` only checks gelsd driver because that is the only option used in NumPy.
`check_correctness_scipy` checks gelsd, gelss, gelsy drivers. Since there is no reference implementation that uses gels driver we check the correctness of the solution ourselves using pseudoinverse in `check_solution_correctness` function.

mruberry commented on torch/linalg/__init__.py:
```diff
@@ -472,7 +473,7 @@
 
 .. warning::
-    The default value for :attr:`cond` is subject to a potential change.
+    The default value for :attr:`rcond` is subject to a potential change.
```
"... is subject to a potential change" -> "... may change in a future PyTorch release."

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk commented:
@mruberry I updated this PR: I refined the conditions for the driver-specific `rcond` value in tests and modified the warning in the documentation about `rcond`.

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

[ghstack-poisoned]

mruberry reviewed (approved):
Nice consistency improvement!

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

Differential Revision: [D27993741](https://our.internmc.facebook.com/intern/diff/D27993741)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

Differential Revision: [D27993741](https://our.internmc.facebook.com/intern/diff/D27993741)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

Differential Revision: [D27993741](https://our.internmc.facebook.com/intern/diff/D27993741)

[ghstack-poisoned]

IvanYashchuk committed:
Update on "Updated linalg.lstsq with NumPy compatible kwarg rcond"


Renamed "cond" -> "rcond" to be NumPy compatible. The default value for
rcond was changed to match non-legacy NumPy behavior.

Differential Revision: [D27993741](https://our.internmc.facebook.com/intern/diff/D27993741)

[ghstack-poisoned]
